icicle-source/src/Icicle/Source/ToCore/ToCore.hs

```diff
@@ -593,3 +602,99 @@ convertReduce xx
  -- so it must be an application of a non-primitive
  | otherwise
  = convertError $ ConvertErrorExpApplicationOfNonPrimitive (annAnnot $ annotOfExp xx) xx
+
+-- | Incorporate the refutation fact key into the query.
+--
+-- @
+--   INIT:
+--     (none, <fold_init>)
+--   KONS:
+--     new_key = nub_exp value
+--     fold_bool
+--       (old_key, acc)
+--       (new_key, <fold_kons>)
+--       (some new_key == old_key)
+-- @
+--
+convertKey :: (Hashable n, Eq n)
+           => ConcreteKey (Annot a n) n
+           -> CoreBinds () n
+           -> ConvertM a n (CoreBinds () n)
+convertKey (ConcreteKey Nothing)  core = return core
+convertKey (ConcreteKey (Just k)) core = do
+  let a'k    = annotOfExp k
+  k'        <- convertExp k
+  t'k       <- convertValType (annAnnot a'k) (annResult a'k)
+  let t'key  = T.OptionT t'k
+
+  let pairX t u = CE.xPrim . C.PrimMinimal . Min.PrimConst    $ Min.PrimConstPair  t u
+  let fstX  t u = CE.xPrim . C.PrimMinimal . Min.PrimPair     $ Min.PrimPairFst    t u
+  let sndX  t u = CE.xPrim . C.PrimMinimal . Min.PrimPair     $ Min.PrimPairSnd    t u
+  let someX t   = CE.xPrim . C.PrimMinimal . Min.PrimConst    $ Min.PrimConstSome  t
+  let eqX   t   = CE.xPrim . C.PrimMinimal $ Min.PrimRelation   Min.PrimRelationEq t
+
+  let nubKons t'acc n'input xx = do
+        n'acc'old <- lift fresh
+        n'acc'new <- lift fresh
+        n'key'old <- lift fresh
+        n'key'new <- lift fresh
+
+        xx' <- lift $ CE.subst1 () n'input (CE.xVar n'acc'old) xx
+        let a  =                    CE.makeApps () (sndX  t'key t'acc) [CE.xVar n'input]
+        let x' = CE.xLet n'key'old (CE.makeApps () (fstX  t'key t'acc) [CE.xVar n'input])
+               $ CE.xLet n'acc'old a
+               $ CE.xLet n'key'new (CE.makeApps () (someX t'k        ) [k'             ])
+               $ CE.xLet n'acc'new xx'
+               $ CE.makeApps ()
+                   ( CE.xPrim (C.PrimFold C.PrimFoldBool (T.PairT t'key t'acc)) )
+                   [ CE.xVar n'input
+                   , CE.makeApps () (pairX t'key t'acc) [ CE.xVar n'key'new, CE.xVar n'acc'new ]
+                   , CE.makeApps () (eqX   t'key      ) [ CE.xVar n'key'old, CE.xVar n'key'new ] ]
+
+        pure ( x', Map.singleton n'input a )
```

Comment:
I was imagining you could add this new_key and fold_bool as two extra folds right at the start, then just change each fold to refer to the fold_bool. rather than adding these to each fold?